### PR TITLE
[Do not merge] Player listener cleanup

### PR DESCRIPTION
**Pull request by ljacqu**
This PR does the following:
- moves the logic of join checks into a separate class
  - create unit tests for most methods
- structures the checks we perform into individual methods
  - offers clean overview of which checks are being performed where, in which order
  - reduces duplication in onPlayerLogin and onPreLogin
- **(!)** moves checks only done in onPreLogin to onPlayerLogin: not all servers fire onPreLogin in offline mode

TODO:
- verify that all features work (name validation, kick non registered, anti bot, name casing, single session, player country) on Spigot and CraftBukkit

Request for comments—@sgdc3 @games647 @DNx5 & anyone interested. I will not merge before Wednesday so feel free to take the time to look the changes over (much appreciated!).

### CODE REVIEW

**Comment by sgdc3** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           
```


See the complete overview on [Codacy](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/pullRequest?prid=245908&bid=3324701)

File: src/main/java/fr/xephi/authme/listener/AuthMePlayerJoinListener.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        final Player player = event.getPlayer();
+        if (player == null || Utils.isUnrestricted(player)) {
+            return;
+        }
+
+        if (event.getResult() == PlayerLoginEvent.Result.KICK_FULL) {
+            if (permissionsManager.hasPermission(player, PlayerStatePermission.IS_VIP)) {
+                int playersOnline = bukkitService.getOnlinePlayers().size();
+                if (playersOnline > plugin.getServer().getMaxPlayers()) {
+                    event.allow();
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/listener/AuthMePlayerJoinListener.java]: Pretty sure the check was wrong; subsequent commit changes `playersOnline > maxPlayers` to `playersOnline < maxPlayers`.

Comment (ljacqu) [inline src/main/java/fr/xephi/authme/listener/AuthMePlayerJoinListener.java]: Pretty sure the check was wrong; subsequent commit changes `playersOnline > maxPlayers` to `playersOnline < maxPlayers`.

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+        // expect
+        expectValidationExceptionWith(MessageKey.MUST_REGISTER_MESSAGE);
+
+        // when
+        onJoinVerifier.checkKickNonRegistered(false);
+    }
+
+    @Test
+    public void shouldNotKickRegisteredPlayer() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l182)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+        // when
+        onJoinVerifier.checkNameCasing(player, auth);
+
+        // then
+        verifyZeroInteractions(dataSource);
+    }
+
+    @Test
+    public void shouldAcceptNameForUnregisteredAccount() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l328)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+
+        // when
+        onJoinVerifier.checkNameCasing(player, auth);
+
+        // then
+        verifyZeroInteractions(dataSource);
+    }
+
+    @Test
+    public void shouldRejectNameWithWrongCasing() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l271)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    public void shouldNotKickRegisteredPlayer() throws FailedVerificationException {
+        // given
+        given(settings.getProperty(RestrictionSettings.KICK_NON_REGISTERED)).willReturn(true);
+
+        // when
+        onJoinVerifier.checkKickNonRegistered(true);
+    }
+
+    @Test
+    public void shouldNotKickUnregisteredPlayer() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l191)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+
+        // when
+        onJoinVerifier.checkNameCasing(player, auth);
+
+        // then
+        verifyZeroInteractions(dataSource);
+    }
+
+    @Test
+    public void shouldAcceptNameThatIsNotOnline() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l342)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+        given(settings.getProperty(RestrictionSettings.MAX_NICKNAME_LENGTH)).willReturn(8);
+        given(settings.getProperty(RestrictionSettings.ALLOWED_NICKNAME_CHARACTERS)).willReturn("[a-zA-Z0-9]+");
+        onJoinVerifier.reload(); // @PostConstruct method
+
+        // when
+        onJoinVerifier.checkIsValidName("Bobby5");
+    }
+
+    @Test
+    public void shouldRejectTooLongName() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l212)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+
+        // expect
+        expectValidationExceptionWith(MessageKey.INVALID_NAME_CHARACTERS, "[a-zA-Z0-9]+");
+
+        // when
+        onJoinVerifier.checkIsValidName("Tester!");
+    }
+
+    @Test
+    public void shouldAllowProperlyCasedName() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l257)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+
+        // expect
+        expectValidationExceptionWith(MessageKey.INVALID_NAME_LENGTH);
+
+        // when
+        onJoinVerifier.checkIsValidName("longerthaneight");
+    }
+
+    @Test
+    public void shouldRejectTooShortName() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l227)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+
+        // when
+        onJoinVerifier.checkNameCasing(player, auth);
+
+        // then
+        verify(dataSource).updateRealName("someone", "SOMEONE");
+    }
+
+    @Test
+    public void shouldAcceptCasingMismatchForDisabledSetting() throws FailedVerificationException {
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l314)

File: src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+    @Mock
+    private DataSource dataSource;
+    @Mock
+    private Messages messages;
+    @Mock
+    private PermissionsManager permissionsManager;
+    @Mock
+    private AntiBot antiBot;
+    @Mock
+    private ValidationService validationService;
```
Comment (sgdc3) [inline src/test/java/fr/xephi/authme/listener/OnJoinVerifierTest.java]: Codacy Issue found: [Avoid unused private fields such as 'validationService'.](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/file/2651420938/issues/source?bid=3316370&fileBranchId=3324701#l66)

**Comment by sgdc3** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           
```


See the complete overview on [Codacy](https://www.codacy.com/app/sgdc3-mail/AuthMeReloaded/pullRequest?prid=245908&bid=3324701)

**Comment by sgdc3** [comment]
Lol codacy is owning me xD

**Comment by sgdc3** [comment]
i think we sould merge this.

**Comment by sgdc3** [comment]
@ljacqu

**Comment by ljacqu** [comment]
@sgdc3 I'll do one last test run and I'll merge it. Deal?

**Comment by sgdc3** [comment]
Wait, one issue

**Comment by sgdc3** [comment]
I think we can optimise the player kick event handler: when someone is kicked the server fires both kick and quit event!

**Comment by sgdc3** [comment]
@ljacqu

**Comment by ljacqu** [comment]
You mean because `onPlayerKick` and `onQuit` will be triggered?

**Comment by sgdc3** [comment]
@ljacqu Exactly

File: src/main/java/fr/xephi/authme/listener/AuthMeEntityListener.java [hunk 12]
```diff
@@ -90,7 +91,7 @@ public void onLowestEntityInteract(EntityInteractEvent event) {
         }
     }
 
-    // In old versions of the Bukkit API getShooter() returns a Player Object instead of a ProjectileSource
+    // TODO #733: Player can't throw snowball but the item is taken.
```
Comment (sgdc3) [inline src/main/java/fr/xephi/authme/listener/AuthMeEntityListener.java]: whoops xD

**Pull request merged**